### PR TITLE
Allowing multiple pages with the same title

### PR DESCRIPTION
**Pull request by finkedave**
Adding an option to confluence_publish_postfix for making it dynamic and thus allowing users to be able to publish pages with the same titles in sub directories

Resolved ticket: #676 

@jdknight Thanks for looking at the ticket and giving the concept your thoughts. Totally good ideas on how to actually calculate the hash. Since the key piece is that it has to be the same regardless of who runs it.

I am going to open this PR to show how I originally have it working on our system. That whole confluence_title_hash_root_path PR that i closed was.a total mess and never meant to open a PR with it. This way is so much cleaner since i just expand on your usage of postfix

But again I am open to all of your suggestions for changing this if there is a possibility of getting this feature into the base repos.

### CODE REVIEW

**Review by jdknight — CHANGES_REQUESTED** [review]
Thanks for the pull request. I do like this idea of utilizing the existing postfix implementation to help apply a generated hash value. I have some recommendations on this approach, if you are willing.

The proposed changes would have a project define the following to utilize hash generation for titles:

```
confluence_publish_postfix = '&unique_hash'
```

Which could generate documents such as `My Title -SF6F5AG`. This has two limitations:

- A user does not have the flexibility to configure how many hash characters to accept.
- Cannot format the trailing hash value to another pattern (e.g. `My Title (SF6F5AG)`).

I wonder if it would be better if we could make this more powerful by allowing a user to format the postfix to decide if they wish to accept a hash value. Consider the following snippet:

```
hash='da39a3ee5e6b4b0d3255bfef95601890afd80709'[:10]
fmt = "My postfix"
print(fmt.format(hash=hash)) # outputs: My postfix
fmt = "My postfix ({hash})"
print(fmt.format(hash=hash)) # outputs: My postfix (da39a3ee5e)
fmt = "My postfix ({hash:.5})"
print(fmt.format(hash=hash)) # outputs: My postfix (da39a)
```

And maybe in the future, additional options can be added into the format string if a user wished to take advantage of them.

I have also adding a series of specific change requests/comments, if you are willing.

**Comment by jdknight** [comment]
In addition, I forgot to mention that the original commit was not signed with a DCO. I would recommend [reading how to address this](https://github.com/sphinx-contrib/confluencebuilder/pull/680/checks?check_run_id=7349336683), since it is a requirement for changes to be pulled into this extension.

**Comment by jdknight** [comment]
Note that there may be a possibility of shipping out a v1.9 release of this extension next weekend (2022-07-[30-31]). If this is a feature you would like in the upcoming release, please let me know and we can try to ensure we can review/fix things up to get this merged in. Note that it is important that the changes are rebased with the DCO set, or I can neither merge this in or try to cleanup any of the proposed changes (if you wanted to have a maintainer finish the work). Otherwise, we can wait for the next following release to try to get this feature in.

**Comment by finkedave** [comment]
> Note that there may be a possibility of shipping out a v1.9 release of this extension next weekend (2022-07-[30-31]). If this is a feature you would like in the upcoming release, please let me know and we can try to ensure we can review/fix things up to get this merged in. Note that it is important that the changes are rebased with the DCO set, or I can neither merge this in or try to cleanup any of the proposed changes (if you wanted to have a maintainer finish the work). Otherwise, we can wait for the next following release to try to get this feature in.

Sorry was on vacation. I wasn't able to find the time to make the changes yet, going to try to spend some time this week on them. All your suggestions are spot on, I just need to find the time to do it.   No rush on getting this into the next release, but I will make sure I have all this done by next week some time so I can get it onto the following release.

**Comment by finkedave** [comment]
I changed the unit testing and methods a bit to make it easier for others to add additional placeholders. Great comments on making this more robust.

File: sphinxcontrib/confluencebuilder/state.py
Comment (jdknight) [inline sphinxcontrib/confluencebuilder/state.py:230]: Out of curiosity, does the use of regular expressions here help that much? I assume it was added to help avoid attempts to invoke a format operation (e.g. generating hashes) if no keywords were detected. I would have been happy if we just performed a format request on a non-empty string (even if not keywords were provided), but if the use of `re` helps here, works for me.

That being said, this implementation is calling `match` per title being processed. I would imagine if you are looking for performance here, we could change this to check the postfix once for any possible keywords. And if any are detected, set a state/class flag to be check on each `_format_postfix` call (also, ensuring the flag is cleared in the `reset()` call). Although, I am not worried about it. If you only want to use the implementation as is, that is fine by me.

Lastly, I believe `search` may be a faster alternative to `match` for the small string literals a postfix value could be (based on some casual browsing on the two types).

File: doc/configuration.rst
Comment (jdknight) [inline doc/configuration.rst:630]: > [Pages] with the same title must have different parent pages.

I do not believe this note is applicable anymore (correct me if I am wrong here).

File: doc/configuration.rst
Comment (jdknight) [inline doc/configuration.rst:624]: While this text is nicely formatted for the raw content, this will not format as expected in the rendered documentation. The trailing lines for the bullet point need to be aligned with the first line's leading content. In addition, it may be nice to keyword the `{hash}` value (if desired). For example:

```
    * ``{hash}`` - Create a reproduceable hash given the title and location based
      from the project root. Using this placeholder provides an option
      ...
```

File: doc/configuration.rst
Comment (jdknight) [inline doc/configuration.rst:622]: Maybe end this with a colon? Or just:

```
    Supported placeholders:
```

File: doc/configuration.rst
Comment (jdknight) [inline doc/configuration.rst:627]: `confluence` -> `Confluence`

File: tests/unit-tests/datasets/postfix_formatting/index.rst
Comment (jdknight) [inline tests/unit-tests/datasets/postfix_formatting/index.rst:7]: I assume there should be a newline here between the directive definition and its contents.

File: tests/unit-tests/datasets/postfix_formatting/index.rst
Comment (jdknight) [inline tests/unit-tests/datasets/postfix_formatting/index.rst:1]: Since this is an index page with a toctree, there is no need to add an `:orphan:` metadata (the same for the child documents that are explicitly referenced inside the `toctree` directive.

File: tests/unit-tests/test_config_postfix_formatting.py
Comment (jdknight) [inline tests/unit-tests/test_config_postfix_formatting.py:14]: Note that we still support Python 2.7, so the mock module will not be available.

**Review by jdknight — CHANGES_REQUESTED** [review]
Thanks for the updates, the changes are looking good to far. Going to stack on some additional comments based off the new change set.

I'll also trigger the runners.

File: sphinxcontrib/confluencebuilder/state.py
Comment (finkedave) [reply sphinxcontrib/confluencebuilder/state.py:230]: I'm good with keeping this file simple. But ya my reasoning is if we keep adding more replaces placeholders I didn't want to make people using static suffixes builds continually slower. I'll remove all the regex stuff to keep the change straightforward then in the future if more placeholders are added we always could add regex back in and only calculate things that have placeholders present. ie don't calculate hash unless {hash} was found in the postfix. So it would be a dynamically called format

File: tests/unit-tests/test_config_postfix_formatting.py
Comment (finkedave) [reply tests/unit-tests/test_config_postfix_formatting.py]: I added the requirement for mock on python < 3.3. To keep the tests clean. If you would perfer not to use mock for this repos due to the python <3.3 requirement I can alter these tests as well to just use a known hash for the docname its not a big deal for me either way

Comment (jdknight) [reply tests/unit-tests/test_config_postfix_formatting.py]: 👍, I have no problem introducing a `mock` requirement for older Python versions for the test environment.

**Review by jdknight — APPROVED** [review]

**Pull request merged**
